core/src/main/java/io/questdb/cairo/O3PartitionJob.java

```diff
@@ -970,6 +1086,183 @@ private static long getDedupRowsWithAdditionalKeys(
+            int rowGroup,
+            int timestampIndex,
+            int timestampColumnType,
+            long mergeRangeLo,
+            long mergeRangeHi,
+            TableRecordMetadata tableWriterMetadata,
+            long srcOooBatchRowSize,
+            Path path
+    ) {
+        long timestampChunkPtr = file.decodeColumnChunk(rowGroup, timestampIndex, timestampColumnType);
+        long timestampDataPtr = PartitionDecoder.getChunkDataPtr(timestampChunkPtr);
+        long rowGroupRowCount = PartitionDecoder.getRowGroupRowCount(timestampChunkPtr);
+
+        long mergeBatchRowCount = mergeRangeHi - mergeRangeLo + 1;
+        long mergeRowCount = mergeBatchRowCount + rowGroupRowCount;
+
+        final long timestampMergeIndexSize = mergeRowCount * TIMESTAMP_MERGE_ENTRY_BYTES;
+        final long timestampMergeIndexAddr = createMergeIndex(
+                timestampDataPtr,
+                sortedTimestampsAddr,
+                0,
+                rowGroupRowCount - 1,
+                mergeRangeLo,
+                mergeRangeHi,
+                timestampMergeIndexSize
+        );
+
+        ObjList<MemoryCARW> mergedMemory = new ObjList<>();
+        try (PartitionDescriptor partitionDescriptor = new BorrowedMemoryPartitionDescriptor()) {
+            partitionDescriptor.of(tableWriter.getTableToken().getTableName(), mergeRowCount, timestampIndex);
+
+            final int columnCount = tableWriterMetadata.getColumnCount();
+            for (int i = 0; i < columnCount; i++) {
+                int columnType = tableWriterMetadata.getColumnType(i);
+                final String columnName = tableWriterMetadata.getColumnName(i);
+                final int columnId = tableWriterMetadata.getColumnMetadata(i).getWriterIndex();
+
+                if (columnType < 0) {
+                    continue;
+                }
+                final boolean notTheTimestamp = i != timestampIndex;
+                final int columnOffset = getPrimaryColumnIndex(i);
+                final MemoryCR oooMem1 = oooColumns.getQuick(columnOffset);
+                final MemoryCR oooMem2 = oooColumns.getQuick(columnOffset + 1);
+
+                final long columnChunkPtr = notTheTimestamp ? file.decodeColumnChunk(rowGroup, i, columnType) : timestampChunkPtr;
+                final long columnDataPtr = PartitionDecoder.getChunkDataPtr(columnChunkPtr);
+                final long columnAuxPtr = PartitionDecoder.getChunkAuxPtr(columnChunkPtr);
+
+                if (ColumnType.isSymbol(columnType)) {
+
+                    final long srcOooFixAddr = oooMem1.addressOf(0);
+
+                    final MapWriter symbolMapWriter = tableWriter.getSymbolMapWriter(i);
+                    final MemoryR offsetsMem = symbolMapWriter.getSymbolOffsetsMemory();
+                    final MemoryR valuesMem = symbolMapWriter.getSymbolValuesMemory();
+
+                    MemoryCARWImpl dstFixMem = new MemoryCARWImpl(PAGE_SIZE, Integer.MAX_VALUE, MemoryTag.NATIVE_O3);
```

Comment:
Please refactor to re-use the allocated memory java objects and underlying memory

Reply:
refactored to have lifetime of `PartitionDescriptor`